File: java/Nodes at given distance in binary tree.java
class Solution
{
    public static ArrayList<Integer> KDistanceNodes(Node root, int target , int k)
    {
        // return the sorted list of all nodes at k dist
        ArrayList<Integer> list = new ArrayList<>();
        findKDN(root,target,k,list);
        Collections.sort(list);
        return list;
        
    }
    
    static void findKDNdown(Node node, int k, ArrayList<Integer> list){
        if(node==null || k<0) return;
        if(k==0){
            list.add(node.data);
            return;
        }
        
        findKDNdown(node.left,k-1,list);
        findKDNdown(node.right,k-1,list);
    }
    
    static int findKDN(Node node, int target, int k, ArrayList<Integer> list){
        if(node==null) return -1;
        if(node.data==target){
            findKDNdown(node,k,list);
            return 0;
        }
        
        int ld = findKDN(node.left,target,k,list);
        if(ld!=-1){
            if(ld+1==k) list.add(node.data);
            else findKDNdown(node.right,k-ld-2,list);
            return ld+1;
        }
        
        int rd = findKDN(node.right,target,k,list);
        if(rd!=-1){
            if(rd+1==k) list.add(node.data);
            else findKDNdown(node.left,k-rd-2,list);
            return rd+1;
        }
        return -1;
    }
    
    
};

File: cpp/Nodes at given distance in binary tree.cpp
class Solution{
public:
    void dfs(Node* root, unordered_map<Node*, Node*>& parent){
        if(root == NULL) return;
        if(root->left){
            parent[root->left] = root;
            dfs(root->left, parent);
        }
        if(root->right){
            parent[root->right] = root;
            dfs(root->right, parent);
        }
    }
    
    void findNode(Node* root, Node* &targetNode, int t){
        if(root == NULL) return;
        if(root->data == t){
            targetNode = root;
            return;
        }
        if(root->left) findNode(root->left, targetNode, t);
        if(root->right) findNode(root->right, targetNode, t);
    }
    
    vector<int> KDistanceNodes(Node* root, int target, int k){
        Node* targetNode;
        findNode(root, targetNode, target);
        
        queue<Node*> q;
        q.push(targetNode);
        
        unordered_map<Node*, Node*> parent;
        parent[root] = NULL;
        dfs(root, parent);
        
        unordered_map<Node*, bool> vis;
        vis[targetNode] = true;
        
        int level = 0;
        while(!q.empty()){
            if(level == k) break;
            level += 1;
            
            int size = q.size();
            
            for(int i=1; i<=size; i++){
                Node* node = q.front();
                q.pop();
                
                if(node->left && !vis[node->left]){
                    q.push(node->left);
                    vis[node->left] = 1;
                }
                
                if(node->right && !vis[node->right]){
                    q.push(node->right);
                    vis[node->right] = 1;
                }
                
                if(parent[node] && !vis[parent[node]]){
                    q.push(parent[node]);
                    vis[parent[node]] = 1;
                }
            }
        }
        
        vector<int> res;
        while(!q.empty()){
            res.push_back(q.front()->data);
            q.pop();
        }
        sort(res.begin(), res.end());
        
        return res;
    }
};